File: backend/app/models/integration.py
"""
Integration model for external services
"""

from sqlalchemy import Column, String, DateTime, Boolean, JSON, ForeignKey, Enum as SQLEnum, Integer
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
import enum

from ..core.database import Base


class IntegrationType(str, enum.Enum):
    """Type of integration"""
    CALDAV = "caldav"
    GOOGLE_CALENDAR = "google_calendar"
    OUTLOOK_CALENDAR = "outlook_calendar"


class SyncDirection(str, enum.Enum):
    """Direction of synchronization"""
    TO_CALENDAR = "to_calendar"  # Workmate → External
    FROM_CALENDAR = "from_calendar"  # External → Workmate
    BIDIRECTIONAL = "bidirectional"  # Both ways


class Integration(Base):
    """External service integrations (calendars, etc.)"""

    __tablename__ = "integrations"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    # Integration details
    name = Column(String, nullable=False)  # User-defined name
    integration_type = Column(SQLEnum(IntegrationType), nullable=False)
    enabled = Column(Boolean, default=True)

    # Configuration (encrypted credentials, URLs, etc.)
    config = Column(JSON, nullable=False)  # Stores URL, username, calendar_id, etc.
    credentials = Column(JSON, nullable=True)  # Encrypted OAuth tokens or passwords

    # Sync settings
    sync_direction = Column(SQLEnum(SyncDirection), default=SyncDirection.BIDIRECTIONAL)
    auto_sync = Column(Boolean, default=True)
    sync_interval_minutes = Column(Integer, default=15)  # How often to sync

    # Sync status
    sync_status = Column(String, default="idle")  # idle, syncing, success, error
    last_sync_at = Column(DateTime, nullable=True)
    error_log = Column(String, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    user = relationship("User", back_populates="integrations")
    calendar_events = relationship("CalendarEvent", back_populates="integration", cascade="all, delete-orphan")

File: frontend/lib/models/integration.dart
class Integration {
  final String id;
  final String userId;
  final String name;
  final String integrationType;
  final bool enabled;
  final Map<String, dynamic> config;
  final String syncDirection;
  final bool autoSync;
  final int syncIntervalMinutes;
  final String syncStatus;
  final DateTime? lastSyncAt;
  final String? errorLog;
  final DateTime createdAt;
  final DateTime updatedAt;

  Integration({
    required this.id,
    required this.userId,
    required this.name,
    required this.integrationType,
    this.enabled = true,
    required this.config,
    this.syncDirection = 'bidirectional',
    this.autoSync = true,
    this.syncIntervalMinutes = 15,
    this.syncStatus = 'idle',
    this.lastSyncAt,
    this.errorLog,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Integration.fromJson(Map<String, dynamic> json) {
    return Integration(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      name: json['name'] as String,
      integrationType: json['integration_type'] as String,
      enabled: json['enabled'] as bool? ?? true,
      config: json['config'] as Map<String, dynamic>,
      syncDirection: json['sync_direction'] as String? ?? 'bidirectional',
      autoSync: json['auto_sync'] as bool? ?? true,
      syncIntervalMinutes: json['sync_interval_minutes'] as int? ?? 15,
      syncStatus: json['sync_status'] as String? ?? 'idle',
      lastSyncAt: json['last_sync_at'] != null
          ? DateTime.parse(json['last_sync_at'])
          : null,
      errorLog: json['error_log'] as String?,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'name': name,
      'integration_type': integrationType,
      'enabled': enabled,
      'config': config,
      'sync_direction': syncDirection,
      'auto_sync': autoSync,
      'sync_interval_minutes': syncIntervalMinutes,
      'sync_status': syncStatus,
      'last_sync_at': lastSyncAt?.toIso8601String(),
      'error_log': errorLog,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  String get typeDisplayName {
    switch (integrationType) {
      case 'caldav':
        return 'CalDAV';
      case 'google_calendar':
        return 'Google Calendar';
      case 'outlook_calendar':
        return 'Outlook Calendar';
      default:
        return integrationType;
    }
  }

  String get syncDirectionDisplayName {
    switch (syncDirection) {
      case 'to_calendar':
        return 'Nur zu Kalender';
      case 'from_calendar':
        return 'Nur vom Kalender';
      case 'bidirectional':
        return 'Bidirektional';
      default:
        return syncDirection;
    }
  }

  bool get hasError => errorLog != null && errorLog!.isNotEmpty;

  Integration copyWith({
    String? name,
    bool? enabled,
    String? syncDirection,
    bool? autoSync,
    int? syncIntervalMinutes,
    Map<String, dynamic>? config,
  }) {
    return Integration(
      id: id,
      userId: userId,
      name: name ?? this.name,
      integrationType: integrationType,
      enabled: enabled ?? this.enabled,
      config: config ?? this.config,
      syncDirection: syncDirection ?? this.syncDirection,
      autoSync: autoSync ?? this.autoSync,
      syncIntervalMinutes: syncIntervalMinutes ?? this.syncIntervalMinutes,
      syncStatus: syncStatus,
      lastSyncAt: lastSyncAt,
      errorLog: errorLog,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
    );
  }
}
